### PR TITLE
Event log

### LINKED ISSUE #216: Event log
* Define the event log format
 * Define combinators
 * Use event logging at errors, redirection, login, logout

### PR DESCRIPTION
**Pull request by fisx**
Fixes #216

### CODE REVIEW

File: tests/Persistent/ApiSpec.hs [hunk 1]
```diff
@@ -100,7 +102,12 @@ addDbSpecProp imp name getXs addX propX =
 
 addDbSpec :: (Foldable f, Arbitrary proto, HasAUpdate ev a)
           => PersistenceImpl -> String -> Query (f a) -> (EnvWith proto -> ev) -> Spec
-addDbSpec imp name getXs addX = addDbSpecProp imp name getXs addX (\_ _ -> passes)
+addDbSpec imp name getXs addX = addDbSpecProp imp name getXs addX id (\_ _ -> passes)
+
+-- | like 'addDbSpec', but takes a function that strips extra info from the translates result value.
```
Comment (andorp) [inline tests/Persistent/ApiSpec.hs]: Probably typo `translates`?

File: src/Persistent/Pure.hs [hunk 2]
```diff
@@ -542,7 +556,7 @@ moveIdeaToTopic ideaId mTopicId =
 setTopicPhase :: AUID Topic -> Phase -> AUpdate ()
 setTopicPhase tid phase = withTopic tid . topicPhase .= phase
 
-addTopic :: Timestamp -> AddDb Topic
```
Comment (fisx) [inline src/Persistent/Pure.hs]: i'll add a comment here explaining things.

Comment (fisx) [inline src/Persistent/Pure.hs]: Also, this should keep the same type, and we should add a function `addTopicYieldLocs`.  That will make lots of the follow-up changes unnecessary.

Thanks, @andorp for the idea!

**Comment by andorp** [comment]
LGTM

**Pull request merged**